streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -1996,6 +2012,128 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
   def done[E, A](exit: Exit[E, A]): ZStream[Any, E, A] =
     fromEffect(ZIO.done(exit))
 
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The optionality of the error type `E` can be used to signal the end of the stream,
+   * by setting it to `None`.
+   */
+  def effectAsync[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Unit) => Unit,
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    effectAsyncMaybe(callback => {
+      register(callback)
+      None
+    }, outputBuffer)
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The registration of the callback returns either a canceler or synchronously returns a stream.
+   * The optionality of the error type `E` can be used to signal the end of the stream, by
+   * setting it to `None`.
+   */
+  def effectAsyncInterrupt[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Unit) => Either[Canceler[R], ZStream[R, E, A]],
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    ZStream {
+      for {
+        output  <- Queue.bounded[Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        eitherStream <- ZManaged.effectTotal {
+                         register(k =>
+                           try {
+                             runtime.unsafeRun(Take.fromPull(k).flatMap(output.offer))
+                             ()
+                           } catch {
+                             case FiberFailure(Cause.Interrupt(_)) =>
+                           }
+                         )
+                       }
+        pull <- eitherStream match {
+                 case Left(canceler) =>
+                   (for {
+                     done <- ZRef.makeManaged(false)
+                   } yield done.get.flatMap {
+                     if (_) Pull.end
+                     else
+                       output.take.flatMap(_.toPull).onError(_ => done.set(true) *> output.shutdown)
+                   }).ensuring(canceler)
+                 case Right(stream) => output.shutdown.toManaged_ *> stream.process
+               }
+      } yield pull
+    }
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times
+   * The registration of the callback itself returns an effect. The optionality of the
+   * error type `E` can be used to signal the end of the stream, by setting it to `None`.
+   */
+  def effectAsyncM[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Unit) => ZIO[R, Nothing, Any],
```

Comment:
Right, you can work around the `ZManaged[R, Nothing` problem and revert this back to `ZIO[R, E, Any]` by formulating this as:
```scala
managed {
  // everything inside ZStream { ...}
}.flatMap(repeatEffectChunkOption(_))
```

Reply:
Done